cluster/query_executor.go

```diff
@@ -883,6 +930,147 @@ type IntoWriteRequest struct {
+	return &remoteIteratorCreator{
+		dialer:  dialer,
+		nodeID:  nodeID,
+		shardID: shardID,
+	}
+}
+
+// CreateIterator creates a remote streaming iterator.
+func (ic *remoteIteratorCreator) CreateIterator(opt influxql.IteratorOptions) (influxql.Iterator, error) {
+	conn, err := ic.dialer.DialNode(ic.nodeID)
+	if err != nil {
+		return nil, err
+	}
+
+	if err := func() error {
+		// Write request.
+		if err := EncodeTLV(conn, createIteratorRequestMessage, &CreateIteratorRequest{
+			ShardID: ic.shardID,
+			Opt:     opt,
+		}); err != nil {
+			return err
+		}
+
+		// Read the response.
+		var resp CreateIteratorResponse
+		if _, err := DecodeTLV(conn, &resp); err != nil {
+			return err
+		} else if resp.Err != nil {
+			return err
+		}
+
+		return nil
+	}(); err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	return influxql.NewReaderIterator(conn)
+}
+
+// FieldDimensions returns the unique fields and dimensions across a list of sources.
+func (ic *remoteIteratorCreator) FieldDimensions(sources influxql.Sources) (fields, dimensions map[string]struct{}, err error) {
+	conn, err := ic.dialer.DialNode(ic.nodeID)
+	if err != nil {
+		return nil, nil, err
+	}
+	defer conn.Close()
+
+	// Write request.
+	if err := EncodeTLV(conn, fieldDimensionsRequestMessage, &FieldDimensionsRequest{
+		ShardID: ic.shardID,
+		Sources: sources,
+	}); err != nil {
+		return nil, nil, err
+	}
+
+	// Read the response.
+	var resp FieldDimensionsResponse
+	if _, err := DecodeTLV(conn, &resp); err != nil {
+		return nil, nil, err
+	}
+	return resp.Fields, resp.Dimensions, resp.Err
+}
+
+// SeriesKeys returns a list of series keys from the underlying shard.
+func (ic *remoteIteratorCreator) SeriesKeys(opt influxql.IteratorOptions) (influxql.SeriesList, error) {
+	conn, err := ic.dialer.DialNode(ic.nodeID)
+	if err != nil {
+		return nil, err
+	}
+	defer conn.Close()
+
+	// Write request.
+	if err := EncodeTLV(conn, seriesKeysRequestMessage, &SeriesKeysRequest{
+		ShardID: ic.shardID,
+		Opt:     opt,
+	}); err != nil {
+		return nil, err
+	}
+
+	// Read the response.
+	var resp SeriesKeysResponse
+	if _, err := DecodeTLV(conn, &resp); err != nil {
+		return nil, err
+	}
+	return resp.SeriesList, resp.Err
+}
+
+// NodeDialer dials connections to a given node.
+type NodeDialer struct {
+	MetaClient MetaClient
+	Timeout    time.Duration
+}
+
+// DialNode returns a connection to a node.
+func (d *NodeDialer) DialNode(nodeID uint64) (net.Conn, error) {
+	ni, err := d.MetaClient.DataNode(nodeID)
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := net.Dial("tcp", ni.TCPHost)
+	if err != nil {
+		return nil, err
+	}
+	conn.SetDeadline(time.Now().Add(d.Timeout))
+
+	// Write the cluster multiplexing header byte
+	if _, err := conn.Write([]byte{MuxHeader}); err != nil {
+		conn.Close()
+		return nil, err
+	}
+
+	return conn, nil
+}
+
+// TSDBStore is an interface for accessing the time series data store.
+type TSDBStore interface {
```

Comment:
In other parts of the codebase these interfaces tend to be defined inside the top-level type. Any reason for this to be hiding down here?